src/sfincs_bmi.cxx

```diff
@@ -11,239 +11,349 @@
 
 #include <cstring>
 #include <iostream>
+
 #include "sfincs_bmi.hxx"
 
 // sfincs_bmi.f90 implements/exports the following bmi functions:
-extern "C" int initialize(char *c_config_file);
-extern "C" int update(double dt);  // Doesn't seem to update current time
+extern "C" int initialize(const char *c_config_file);
 extern "C" int finalize();
-extern "C" void get_start_time(double *tstart);
-extern "C" void get_end_time(double *tend);
-extern "C" void get_time_step(double *dt);
-extern "C" void get_current_time(double *tcurrent);
-extern "C" void get_var_type(const char *c_var_name, char *c_type);
+
+extern "C" int get_component_name(char *name);
+
+extern "C" int update();
+extern "C" int update_until(double t);
+
+extern "C" int get_start_time(double *tstart);
+extern "C" int get_end_time(double *tend);
+extern "C" int get_time_step(double *dt);
+extern "C" int get_time_units(char *units);
+extern "C" int get_current_time(double *tcurrent);
+
+extern "C" int get_var_type(const char *c_var_name, char *c_type);
 
 // These function are also exported but not in the BMI spec
-extern "C" void get_var(const char *c_var_name, void *x);  // not exported; should be get_value?
-extern "C" void set_var(const char *c_var_name, float *xptr);  // should be set get_value?
-extern "C" void get_var_shape(const char *c_var_name, int *var_shape); // should be get_grid_shape
-// extern "C" void get_var_rank(char *c_var_name, int *rank); // should be get_grid_rank
+extern "C" int get_value(const char *c_var_name, void *x); // not exported; should be get_value?
+extern "C" int get_value_at_indices(const char *name, void *dest, int *inds, int count);
+extern "C" int get_value_ptr(const char *name, void **dest_ptr);
+
+extern "C" int set_value(const char *c_var_name, float *xptr); // should be set get_value?
+extern "C" int set_value_at_indices(const char *name, int *inds, int count, void *src);
+
+extern "C" int get_var_shape(const char *c_var_name, int *var_shape); // should be get_grid_shape
+// extern "C" int get_var_rank(char *c_var_name, int *rank); // should be get_grid_rank
 
 // The get_grid_* functions should accept grid id as first argument
-extern "C" void get_grid_type(char *c_type);
-extern "C" void get_grid_rank(int *rank);
-extern "C" void get_grid_size(int *size);
-extern "C" void get_grid_x(double *x);
-extern "C" void get_grid_y(double *y);
+extern "C" int get_grid_type(char *c_type);
+extern "C" int get_grid_rank(int *rank);
+extern "C" int get_grid_size(int *size);
+extern "C" int get_grid_x(double *x);
+extern "C" int get_grid_y(double *y);
```

Comment:
`get_grid_x` and `get_grid_y` don't work as expected. Suppose `get_grid_shape` returns `[31, 14]`, then `get_grid_x` would return an array of length 14 and `get_grid_y` would return a 1d array of length 31. `get_values` would return a flat array of length `31x14`. 

Currently, trying `get_grid_x` with any trial and error length leads to segfaults.

Reply:
Update: above comment doesn't apply for sfincs data structure. All = flat.